tests/h/views/activity_test.py

```diff
@@ -127,36 +131,37 @@ def fake_has_permission(permission, context=None):
         result = controller.search()
 
         assert 'join.html' in pyramid_request.override_renderer
-        assert result == {'group': group}
+        assert result == {'group': test_group}
 
-    def test_renders_join_template_when_not_logged_in(self, controller, factories, pyramid_request, group):
+    @pytest.mark.parametrize('test_group', GROUP_TYPE_OPTIONS, indirect=['test_group'])
+    def test_renders_join_template_when_not_logged_in(self, controller, factories, pyramid_request, test_group):
         """
         If user is logged out and has no read permission, prompt to login and join.
         """
         def fake_has_permission(permission, context=None):
             if permission in ('read', 'join'):
                 return False
-        pyramid_request.user = None
         pyramid_request.has_permission = mock.Mock(side_effect=fake_has_permission)
 
         result = controller.search()
 
         assert 'join.html' in pyramid_request.override_renderer
-        assert result == {'group': group}
+        assert result == {'group': test_group}
 
-    def test_raises_not_found_when_no_read_or_join_permissions(self, controller, factories, pyramid_request):
+    @pytest.mark.parametrize('test_group,test_user', [('group', 'member')], indirect=['test_group', 'test_user'])
+    def test_raises_not_found_when_no_read_or_join_permissions(self, controller, factories, pyramid_request, test_group, test_user):
         def fake_has_permission(permission, context=None):
             if permission in ('read', 'join'):
                 return False
-        pyramid_request.user = factories.User()  # User logged in
         pyramid_request.has_permission = mock.Mock(side_effect=fake_has_permission)
 
         with pytest.raises(httpexceptions.HTTPNotFound):
             controller.search()
 
+    @pytest.mark.parametrize('test_group', GROUP_TYPE_OPTIONS, indirect=['test_group'])
     def test_search_redirects_if_slug_wrong(self,
```

Comment:
Hmm, there is some overhead for each test that runs, so I would be inclined to be slightly cautious about parametrising every test in this class to use each of the group types. I leave this up to you.